Add a guide for `nnx.bridge` (#4171)
google/flax: docs/nnx/bridge_guide.md

> # Use NNX along with Linen

Author: Done.

> This guide is for existing Flax users who want to make their codebase a mixture of Linen and NNX modules, which is made possible by the `nnx.bridge` API. 

Author: Done.

This will be helpful if you:

* Want to migrate your codebase to NNX gradually, one module at a time;
* Have external dependency that already moved to NNX but you haven't, or is still in Linen while you've moved to NNX.

> We hope this allows you to move and try out NNX at your own pace, and leverage the best of both worlds. We will also talk about how to resolve the coveats of interoping the two APIs, on a few aspects that they are fundamentally different.

Reviewer: ```suggestion
We hope this allows you to move and try out NNX at your own pace, and leverage the best of both worlds. We will also talk about how to resolve the caveats of interoperating the two APIs, on a few aspects that they are fundamentally different.
```

Author: Done!

**Note**: 

This guide is about glueing Linen and NNX modules. To migrate an existing Linen module to NNX, check out the [Migrate from Linen to NNX](https://flax.readthedocs.io/en/latest/nnx/haiku_linen_vs_nnx.html) guide. 

> And all built-in Linen layers should have equivalent NNX versions! Check out the list of [Built-in NNX layers](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/index.html).

Author: Done.

```python
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'

from flax import nnx
from flax import linen as nn
from flax.nnx import bridge
import jax
from jax import numpy as jnp
from jax.experimental import mesh_utils
from typing import *
```

## Submodule is all you need

A Flax model is always a tree of modules - either old Linen modules (`flax.linen.Module`, usually written as `nn.Module`) or NNX modules (`nnx.Module`). 

> An `nnx.bridge` wrapper glues the two types together, in both ways:

Reviewer: ```suggestion
The `nnx.bridge` wrapper glues the two types together, in both ways:
```

* **Lazy vs. eager**: Linen modules only allocate space to create variables when they actually see their input. Whereas NNX module instances create their variables the moment they are instantiated, without seeing a sample input.

With that in mind, let's look at how the `nnx.bridge` wrappers tackle the differences.

### Linen -> NNX

Since Linen modules may require an input to create variables, we semi-formally supported lazy initialization in the NNX modules converted from Linen. The Linen variables are created when you give it a sample input.

For you, it's calling `nnx.bridge.lazy_init()` where you call `module.init()` in Linen code.

```python
class LinenDot(nn.Module):
  out_dim: int
  w_init: Callable[..., Any] = nn.initializers.lecun_normal()
  @nn.compact
  def __call__(self, x):
    # Linen might need the input shape to create the weight!
    w = self.param('w', self.w_init, (x.shape[-1], self.out_dim))
    return x @ w

x = jax.random.normal(jax.random.key(42), (4, 32))

> model = bridge.ToNNX(LinenDot(64), rngs=nnx.Rngs(0))  # => `model = LinenDot(64)` in Linen

Reviewer: It would be interesting to  do an `nnx.display` here after construction to visualize the uninitialized structure, and in a new cell call the rest and visualize the module again after its initialized.

Author: Good idea! Done.

x = jax.random.normal(jax.random.key(42), (4, 32))
model = bridge.lazy_init(NNXOuter(64, rngs=nnx.Rngs(0)), x)  # Can fit them into one line too
w, b = model.dot.params['w'], model.b
print(w.shape, b.shape)
y = model(x)
print(y.shape)
```

    (32, 64) (1, 64)
    (4, 64)

The Linen weight is already converted to a typical NNX variable, which is a thin wrapper of the actual JAX array value within. Here, `w` is an `nnx.Param` because it belongs to the `params` collection of `LinenDot` module.

We will talk more about different collections and types in the [NNX Variable <-> Linen Collections](#variable-types-vs-collections) section. Right now, just know that they are converted to NNX variables like native ones.

```python
assert isinstance(model.dot.params['w'], nnx.Param)
assert isinstance(model.dot.params['w'].value, jax.Array)
```

### NNX -> Linen

To convert an NNX module to Linen, you should forward your creation arguments to `bridge.ToLinen` and let it handle the actual creation process. 

We don't let you create an NNX module of your own because new variables are created eagerly as soon as a NNX module is created. Instead, `bridge.ToLinen` will handle it for you.

> An NNX module instance initializes all its variables eagerly when it is created, which consumes memory and compute. On the other hand, Linen modules are stateless, and the typical `init` and `apply` process involves multiple creation of them. To bypass this issue, you should send your creation arguments (instead of a created NNX module) to `bridge.to_linen`, and let it handle the rest.

Reviewer: There is a bit of repetition here.

Author: Done.

```python
class NNXDot(nnx.Module):
  def __init__(self, in_dim: int, out_dim: int, rngs: nnx.Rngs):
    self.w = nnx.Param(nnx.initializers.lecun_normal()(rngs.params(), (in_dim, out_dim)))
  def __call__(self, x: jax.Array):
    return x @ self.w

x = jax.random.normal(jax.random.key(42), (4, 32))

> model = bridge.ToLinen(NNXDot, args=(32, 64))  # <- Pass in the arguments, not an actual module

Reviewer: We should use `to_linen` as its both more ergonomic and does the `FrozenDict` conversion automatically. Most users will copy the first example they see.

Author: Done.

variables = model.init(jax.random.key(0), x)
y = model.apply(variables, x)

print(list(variables.keys()))
print(variables['params']['w'].value.shape)  # => (32, 64)
print(y.shape)                               # => (4, 64)

```

    ['nnx', 'params']
    (32, 64)
    (4, 64)

Note that `ToLinen` modules need to track an extra variable collection - `nnx` - for the static metadata of the underlying NNX module.

```python
# This new field stores the static data that defines the underlying `NNXDot`
print(type(variables['nnx']['graphdef']))    # => `nnx.graph.NodeDef`
```

    <class 'flax.nnx.nnx.graph.NodeDef'>

> You can use a shortcut `bridge.to_linen` to avoid explicitly grouping args and kwargs separately. You only need to use the underlying `bridge.ToLinen()` in some rare cases.

Reviewer: Maybe we can also forward the other arguments such that you can always use `to_linen`?

```python

> bridge.to_linen(NNXDot, 32, out_dim=64) == bridge.ToLinen(NNXDot, args=(32,), kwargs={'out_dim': 64})

Reviewer: I feel that `to_linen` should be the main API. Lets not show 2 ways of doing this.

Author: Done. Explained a use case where you must use `ToLinen`.

```python
print(type(variables['params']['w']))         # => nnx.bridge.NNXMeta
print(type(variables['params']['w'].value))   # => jax.Array
```

    <class 'flax.nnx.nnx.bridge.variables.NNXMeta'>
    <class 'jaxlib.xla_extension.ArrayImpl'>

Similar to `ToNNX`, you can use `ToLinen` to create a submodule of another Linen module. 

```python
class LinenOuter(nn.Module):
  out_dim: int
  @nn.compact
  def __call__(self, x):
    dot = bridge.to_linen(NNXDot, x.shape[-1], self.out_dim)
    b = self.param('b', nn.initializers.lecun_normal(), (1, self.out_dim))
    return dot(x) + b

x = jax.random.normal(jax.random.key(42), (4, 32))
model = LinenOuter(out_dim=64)
y, variables = model.init_with_output(jax.random.key(0), x)
w, b = variables['params']['ToLinen_0']['w'].value, variables['params']['b']
print(w.shape, b.shape, y.shape)
```

    (32, 64) (1, 64) (4, 64)

## Handling RNG keys

All Flax modules, Linen or NNX, automatically handle the RNG keys for variable creation and random layers like dropouts. However, the specific logics of RNG key splitting are different, so you cannot generate the same params between Linen and NNX modules, even if you pass in same keys.

> Another difference is that NNX modules are stateful, so they can track and update the RNG keys within themselves. This means a bit difference in both `ToNNX` and `ToLinen`:

Reviewer: ```suggestion
Another difference is that NNX modules are stateful, so they can track and update the RNG keys within themselves. This means RNG handling is bit different in both `ToNNX` and `ToLinen`:
```

> Another difference is that NNX modules are stateful, so they can track and update the RNG keys within themselves. This means a bit difference in both `ToNNX` and `ToLinen`:

Reviewer: ```
This means a bit difference in both `ToNNX` and `ToLinen`
```
This part is a bit unclear.

Author: Removed.

### Linen to NNX

If you convert a Linen module to NNX, you enjoy the stateful benefit and don't need to pass in extra RNG keys on every module call. You can use always `nnx.reseed` to reset the RNG state within.

```python
x = jax.random.normal(jax.random.key(42), (4, 32))
model = bridge.ToNNX(nn.Dropout(rate=0.5, deterministic=False), rngs=nnx.Rngs(dropout=0))
bridge.lazy_init(model, x)       # We don't really need this b/c no extra params were created here,
                                 # but it's a good practice to always add this line.
y1, y2 = model(x), model(x)
assert not jnp.allclose(y1, y2)  # Two runs yield different outputs!

# Reset the dropout RNG seed, so that next model run will be the same as the first.
nnx.reseed(model, dropout=0)
assert jnp.allclose(y1, model(x))

> ```

Reviewer: Love this example!

Now, it really depends on whether your underlying NNX module generates new random data from its RNG state, or from the passed-in argument. Fortunately, `nnx.Dropout` supports both - using passed-in keys if there is any, and use its own RNG state if not.

And this leaves you with two style options of handling the RNG keys: 

* The NNX style (recommended): Let the underlying NNX state manage the RNG keys, no need to pass in extra keys in `apply()`. This means a few more lines to mutate the `variables` for every apply call, but things will look easier once your whole model no longer needs `ToLinen`.

* The Linen style: Just pass different RNG keys for every `apply()` call.

```python
x = jax.random.normal(jax.random.key(42), (4, 32))
model = bridge.to_linen(nnx.Dropout, rate=0.5)
variables = model.init({'dropout': jax.random.key(0)}, x)

# The NNX RNG state was stored inside `variables`
print('The RNG key in state:', variables['RngKey']['rngs']['dropout']['key'].value)

> print('How many times we have created new RNG keys from it:', variables['RngCount']['rngs']['dropout']['count'].value)
> 
> # NNX style: Must set `RngCount` as mutable and update the variables after every `apply`
> y1, updates = model.apply(variables, x, mutable=['RngCount'])
> variables |= updates
> y2, updates = model.apply(variables, x, mutable=['RngCount'])
> assert not jnp.allclose(y1, y2)  # Every call yields different output!

Reviewer: ```suggestion
print('Key split count:', variables['RngCount']['rngs']['dropout']['count'].value)

# NNX style: Must set `RngCount` as mutable and update the variables after every `apply`
y1, updates = model.apply(variables, x, mutable=['RngCount'])
variables |= updates
y2, updates = model.apply(variables, x, mutable=['RngCount'])
variables |= updates

print('Key split count:', variables['RngCount']['rngs']['dropout']['count'].value)
assert not jnp.allclose(y1, y2)  # Every call yields different output!
```

Author: Sounds great! Done

# Linen style: Just pass different RNG keys for every `apply()` call.
y3 = model.apply(variables, x, rngs={'dropout': jax.random.key(1)})

> y4 = model.apply(variables, x, rngs={'dropout': jax.random.key(2)})

Reviewer: could be interesting to do the opposite and use the same key to show that this overrides the keys and yields the same output

Author: Good idea! done.

    )
    Param(
      value=Array([0., 0., 0.], dtype=float32)
    )
    counter(
      value=Array(0, dtype=int32)
    )
    <class 'abc.counter'>
    (Intermediate(
      value=Array(6.932987, dtype=float32)
    ),)

You can quickly separate different types of NNX variables apart using `nnx.split`.

This can be handy when you only want to set some variables as trainable.

```python
# Separate variables of different types with nnx.split
CountType = type(model.counter['count'])
static, params, counter, the_rest = nnx.split(model, nnx.Param, CountType, ...)
print('All Params:', list(params['params'].keys()))
print('All Counters:', list(counter['counter'].keys()))
print('All the rest (intermediates and RNG keys):', list(the_rest.keys()))

model = nnx.merge(static, params, counter, the_rest)  # You can merge them back at any time
```

    All Params: ['b', 'w']
    All Counters: ['count']
    All the rest (intermediates and RNG keys): ['intermediates', 'rngs']

### NNX to Linen

If you define custom NNX variable types, you should register their names with `nnx.register_variable_name_type_pair` so that they go to the desired collections.

```python
class Count(nnx.Variable): pass

> nnx.register_variable_name_type_pair('Count', Count, overwrite=True)

Reviewer: `Count` would be the automatic name anyway, maybe we can show registering a Linen-style name (no capital, plural)?
```suggestion
nnx.register_variable_name_type_pair('counts', Count, overwrite=True)
```

Author: Done.

The `bridge.ToNNX` and `bridge.ToLinen` API will automatically convert the sharding annotations, if you use the built-in annotation methods (aka. `nn.with_partitioning` for Linen and `nnx.with_partitioning` for NNX).

### Linen to NNX

Even if you are not using any partition metadata in your Linen module, the variable JAX arrays will be converted to `nnx.Variable`s that wraps the true JAX array within. 

If you use `nn.with_partitioning` to annotate your Linen module's variables, the annotation will be converted to a `.sharding` field in the corresponding `nnx.Variable`. 

You can then use `nnx.with_sharding_constraint` to explicitly put the arrays into the annotated partitions within a `jax.jit`-compiled function, to initialize the whole model with every array at the right sharding.

```python
class LinenDotWithPartitioning(nn.Module):
  out_dim: int
  @nn.compact
  def __call__(self, x):
    w = self.param('w', nn.with_partitioning(nn.initializers.lecun_normal(), ('in', 'out')), 
                   (x.shape[-1], self.out_dim))
    return x @ w

@jax.jit
def create_sharded_nnx_module(x):
  model = bridge.lazy_init(bridge.ToNNX(LinenDotWithPartitioning(64), rngs=nnx.Rngs(0)), x)

>   static, state = nnx.split(model)

Reviewer: We started using `graphdef` instead of `static` as its a bit more descriptive.

> @jax.jit
> def create_sharded_nnx_module(x):
>   model = bridge.lazy_init(bridge.ToNNX(LinenDotWithPartitioning(64), rngs=nnx.Rngs(0)), x)
>   static, state = nnx.split(model)
>   sharded_state = nnx.with_sharding_constraint(state, nnx.get_partition_spec(state))
>   return static, sharded_state

Reviewer: Consider using `nnx.jit` here so you don't have to use `merge` later. 
```suggestion
@nnx.jit
def create_sharded_nnx_module(x):
  model = bridge.lazy_init(bridge.ToNNX(LinenDotWithPartitioning(64), rngs=nnx.Rngs(0)), x)
  state = nnx.state(model)
  sharded_state = nnx.with_sharding_constraint(state, nnx.get_partition_spec(state))
  nnx.update(model, sharded_state)
  return model
```

Author: Done. I should get more used to it!

print(f'We have {len(jax.devices())} fake JAX devices now to partition this model...')
mesh = jax.sharding.Mesh(devices=mesh_utils.create_device_mesh((2, 4)), axis_names=('in', 'out'))
x = jax.random.normal(jax.random.key(42), (4, 32))
with mesh:

>   model = nnx.merge(*create_sharded_nnx_module(x))

Reviewer: (If the previous is implemented)
```suggestion
  model = create_sharded_nnx_module(x)
```

Author: Done.

# Use Flax NNX along with Flax Linen

> This guide is for existing Flax users who want to make their codebase a mixture of Flax Linen and Flax NNX `Module`s, which is made possible thanks to the `flax.nnx.bridge` API. 

Reviewer 2: There are redundant trailing spaces on several lines.

```suggestion
This guide is for existing Flax users who want to make their codebase a mixture of Flax Linen and Flax NNX `Module`s, which is made possible thanks to the `flax.nnx.bridge` API.
```

Author: Thanks Ayaka! It seems that our internal & external code checkers doesn't flag out trailing spaces in notebook markdowns so I got lazy and didn't fix them... Will keep an eye out on the next PR!
